From 1427b16052cdd3df160f066580d91c1dcd05b842 Mon Sep 17 00:00:00 2001
From: Jonathan Maple <jmaple@ciq.com>
Date: Tue, 27 Jan 2026 20:06:17 -0500
Subject: [PATCH] drm/vmwgfx: Validate command header size against
 SVGA_CMD_MAX_DATASIZE

jira VULN-161158
cve CVE-2025-40277
commit-author Ian Forbes <ian.forbes@broadcom.com>
commit 32b415a9dc2c212e809b7ebc2b14bc3fbda2b9af

This data originates from userspace and is used in buffer offset
calculations which could potentially overflow causing an out-of-bounds
access.

Fixes: 8ce75f8ab904 ("drm/vmwgfx: Update device includes for DX device functionality")
	Reported-by: Rohit Keshri <rkeshri@redhat.com>
	Signed-off-by: Ian Forbes <ian.forbes@broadcom.com>
	Reviewed-by: Maaz Mombasawala <maaz.mombasawala@broadcom.com>
	Signed-off-by: Zack Rusin <zack.rusin@broadcom.com>
Link: https://patch.msgid.link/20251021190128.13014-1-ian.forbes@broadcom.com
(cherry picked from commit 32b415a9dc2c212e809b7ebc2b14bc3fbda2b9af)
	Signed-off-by: Jonathan Maple <jmaple@ciq.com>
---
 drivers/gpu/drm/vmwgfx/vmwgfx_execbuf.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/gpu/drm/vmwgfx/vmwgfx_execbuf.c b/drivers/gpu/drm/vmwgfx/vmwgfx_execbuf.c
index 2e52d73eba48..76a477ac1541 100644
--- a/drivers/gpu/drm/vmwgfx/vmwgfx_execbuf.c
+++ b/drivers/gpu/drm/vmwgfx/vmwgfx_execbuf.c
@@ -3679,6 +3679,11 @@ static int vmw_cmd_check(struct vmw_private *dev_priv,
 
 
 	cmd_id = header->id;
+	if (header->size > SVGA_CMD_MAX_DATASIZE) {
+		VMW_DEBUG_USER("SVGA3D command: %d is too big.\n",
+			       cmd_id + SVGA_3D_CMD_BASE);
+		return -E2BIG;
+	}
 	*size = header->size + sizeof(SVGA3dCmdHeader);
 
 	cmd_id -= SVGA_3D_CMD_BASE;
-- 
2.47.3

